tests/tensorbox/test_tensorbox.py

```diff
@@ -0,0 +1,541 @@
+
+        xT = qml.TensorBox(x)
+        res = xT * y
+        assert np.all(res.unbox() == x * y)
+
+        yT = qml.TensorBox(y)
+        res = x * yT
+        assert np.all(res.unbox() == x * y)
+
+        res = xT * yT
+        assert np.all(res.unbox() == x * y)
+
+    def test_unbox_list(self, autograd):
+        """Test unboxing a mixed list works correctly"""
+        np = qml.numpy
+        x = np.array([[1, 2], [3, 4]])
+        y = np.array([[1, 0], [0, 1]])
+
+        xT = qml.TensorBox(x)
+        res = xT.unbox_list([y, xT, x])
+
+        assert np.all(res == [y, x, x])
+
+    def test_numpy(self, autograd):
+        """Test that calling numpy() returns a NumPy array representation
+        of the TensorBox"""
+        x = qml.numpy.array([[1, 2], [3, 4]])
+        xT = qml.TensorBox(x)
+        assert isinstance(xT.numpy(), np.ndarray)
+        assert not isinstance(xT.numpy(), qml.numpy.tensor)
+        assert np.all(xT.numpy() == x)
+
+    def test_shape(self, autograd):
+        """Test that arrays return the right shape"""
+        np = qml.numpy
+        x = np.array([[[1, 2], [3, 4]]])
+        x = qml.TensorBox(x)
+        res = x.shape
+        assert res == (1, 2, 2)
+
+    def test_expand_dims(self, autograd):
+        """Test that dimension expansion works"""
+        from pennylane.tensorbox.autograd_box import AutogradBox
+
+        np = qml.numpy
+        x = np.array([1, 2, 3])
+        xT = qml.TensorBox(x)
+
+        res = xT.expand_dims(axis=1)
+        expected = np.expand_dims(x, axis=1)
+        assert isinstance(res, AutogradBox)
+        assert np.all(res.unbox() == expected)
+
+    def test_ones_like(self, autograd):
+        """Test that all ones arrays are correctly created"""
+        from pennylane.tensorbox.autograd_box import AutogradBox
+
+        np = qml.numpy
+        x = np.array([[1, 2, 3], [4, 5, 6]])
+        xT = qml.TensorBox(x)
+
+        res = xT.ones_like()
+        expected = np.ones_like(x)
+        assert isinstance(res, AutogradBox)
+        assert np.all(res.unbox() == expected)
+
+    def test_stack(self, autograd):
+        """Test that arrays are correctly stacked together"""
+        np = qml.numpy
+        x = np.array([[1, 2], [3, 4]])
+        y = np.array([[1, 0], [0, 1]])
+
+        xT = qml.TensorBox(x)
+        res = xT.stack([y, xT, x])
+
+        assert np.all(res.unbox() == np.stack([y, x, x]))
+
+    def test_transpose(self, autograd):
+        """Test that the transpose is correct"""
+        np = qml.numpy
+        x = np.array([[1, 2], [3, 4]])
+        xT = qml.TensorBox(x)
+
+        assert np.all(xT.T.unbox() == x.T)
+
+    def test_autodifferentiation(self, autograd):
+        """Test that autodifferentiation is preserved when writing
+        a cost function that uses TensorBox method chaining"""
+        np = qml.numpy
+        x = np.array([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
+
+        cost_fn = lambda a: (qml.TensorBox(a).T ** 2).unbox()[0, 1]

```

Comment:
Will a user always have to unbox if they want to do something like this?

Reply:
With the new functional approach, the unboxing is no longer visible